File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/array/difference/P1109_CorporateFlightBookings.java
package com.dsa.leetcode.labuladong.array.difference;

public class P1109_CorporateFlightBookings {
    public int[] corpFlightBookings(int[][] bookings, int n) {
        Difference diff = new Difference(n);
        for (int[] booking : bookings) {
            diff.increment(booking[0] - 1, booking[1] - 1, booking[2]);
        }
        return diff.result();
    }

    static class Difference {
        private int[] diff;

        public Difference(int n) {
            diff = new int[n];
        }

        public void increment(int i, int j, int val) {
            diff[i] += val;
            if (j + 1 < diff.length) {
                diff[j + 1] -= val;
            }
        }

        public int[] result() {
            int n = diff.length;
            int[] ans = new int[n];
            ans[0] = diff[0];
            for (int i = 1; i < n; i++) {
                ans[i] = ans[i - 1] + diff[i];
            }
            return ans;
        }
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/array/difference/P1109_CorporateFlightBookings.py
# author: tobebetter9527
# since: 2023/12/13 9:09
from typing import List


class Solution:
    def corpFlightBookings(self, bookings: List[List[int]], n: int) -> List[int]:
        diff = self.Difference(n)
        for booking in bookings:
            diff.increment(booking[0] - 1, booking[1] - 1, booking[2])
        return diff.result()

    class Difference:
        def __init__(self, n: int) -> None:
            self.diff = [0 for i in range(n)]

        def increment(self, i: int, j: int, val: int):
            self.diff[i] += val
            if j + 1 < len(self.diff):
                self.diff[j + 1] -= val

        def result(self):
            n = len(self.diff)
            nums = [0 for i in range(n)]
            nums[0] = self.diff[0]
            for i in range(1, n):
                nums[i] = nums[i - 1] + self.diff[i]
            return nums